packages/common/src/index-api.ts

```diff
@@ -24,13 +24,75 @@ export type BackwardPagination = {
  */
 export type Pagination = ForwardPagination | BackwardPagination
 
+/**
+ * Boolean field value filter
+ */
+export type BooleanValueFilter = { isNull: boolean } | { equalTo: boolean }
+
+/**
+ * Enum field value filter
+ */
+export type EnumValueFilter =
+  | { isNull: boolean }
+  | { equalTo: string }
+  | { notEqualTo: string }
+  | { in: Array<string> }
+  | { notIn: Array<string> }
+
+/**
+ * String or number field value filter
+ */
+export type ScalarValueFilter<T extends string | number = string | number> =
+  | { isNull: boolean }
+  | { equalTo: T }
+  | { notEqualTo: T }
+  | { in: Array<T> }
+  | { notIn: Array<T> }
+  | { lessThan: T }
+  | { lessThanOrEqualTo: T }
+  | { greaterThan: T }
+  | { greaterThanOrEqualTo: T }
+
+/**
+ * Any supported field value filter on an object
+ */
+export type AnyValueFilter = BooleanValueFilter | EnumValueFilter | ScalarValueFilter
+
+/**
+ * Mapping of object keys to value filter
+ */
+export type ObjectFilter = Record<string, AnyValueFilter>
+
+/**
+ * Advanced query filters on a document fields
+ */
+export type QueryFilters =
+  | { doc: ObjectFilter }
+  | { and: Array<QueryFilters> }
```

Comment:
Specifying `and` and `or ` this way allows creation of invalid query filters, such as `{and: []}` or even worse, queries like `{and: {... {and: []} ... }}`.

I'd like to see a strong type here like `{and: {lhs: QueryFilters, rhs: QueryFilters}}`. We could do something similar to prevent deep queries, but at least this prevents just passing empty arrays.

Reply:
I guess there's no way with the type system to express that an array field must be of length greater than X?

Reply:
You could introduce a type that is an array of length one or more. Can even use https://www.npmjs.com/package/@pacote/non-empty-array